src/wlserver.cpp

```diff
@@ -1083,6 +1084,208 @@ static void create_gamescope_control( void )
 	wl_global_create( wlserver.display, &gamescope_control_interface, version, NULL, gamescope_control_bind );
 }
 
+////////////////////////
+// commit-timing
+////////////////////////
+
+enum timer_stage {
+	TIMER_STAGE_LATCH = 0,
+	TIMER_STAGE_PRESENT = 1
+};
+
+enum timer_rounding_mode {
+	TIMER_ROUNDING_NEAREST = 0,
+	TIMER_ROUNDING_NOT_BEFORE = 1
+};
+
+struct commit_timer {
+	struct wl_resource *resource;
+
+	struct wlr_surface *surface;
+	enum timer_stage curr_stage;
+	enum timer_rounding_mode curr_rounding_mode;
+	uint64_t curr_timestamp_nsec;
+
+	struct wl_listener client_commit;
+	struct wlr_addon surface_addon;
+};
+
+static inline int64_t timespec_to_nsec(const struct timespec *a) {
+	return (int64_t)a->tv_sec * 1'000'000'000ul + a->tv_nsec;
+}
+
+static void resource_handle_destroy(struct wl_client *client, struct wl_resource *resource) {
+       wl_resource_destroy(resource);
+}
+
+static void commit_timer_destroy(struct commit_timer *timer) {
+	if (timer == NULL) {
+		return;
+	}
+	wl_list_remove(&timer->client_commit.link);
+	wlr_addon_finish(&timer->surface_addon);
+	wl_resource_set_user_data(timer->resource, NULL); // make inert
+	free(timer);
+}
+
+static void surface_timer_addon_handle_destroy(struct wlr_addon *addon) {
+	struct commit_timer *timer = wl_container_of(addon, timer, surface_addon);
+	commit_timer_destroy(timer);
+}
+
+static const struct wlr_addon_interface surface_timer_addon_impl = {
+	.name = "wp_commit_timer_v1",
+	.destroy = surface_timer_addon_handle_destroy,
+};
+
+#define NANOS_IN_SEC 1000000000
+static uint64_t timer_get_target_present_nsec(uint64_t timestamp_nsec, uint32_t rounding_mode) {
+	uint64_t refresh_nsec = NANOS_IN_SEC/(GetVBlankTimer().GetRefresh()/1000);
```

Comment:
1. We have utilities to do refresh -> cycle already `mHzToRefreshCycle`

2. Doesn't all this rounding logic need to actually be where we check the timestamp in steamcompmgr?

Right now this problem will only manifest on hotplug, which isn't a big deal -- but some day we will support multiple displays, so we should move it there.

Reply:
(I know that adding stuff to the ResListEntry whatever crap is painful, it is high on my to-do to rework that garbage soon and have wlserver create the commits directly).

Reply:
Hope it's better now.